### PR TITLE
Test cases for _is_lambert and Improvement of it !

### PR DESCRIPTION
**Pull request by MaqOwais**
Tried to make a better version according to the latest updates as well as removing some errors accordingly

#14972,#18759,#20537,#20545, #20613

* solveset
  * improvement of  _is_lambert
  * extended to trigonometric and hyperbolic functions
  * test cases

NO ENTRY

### CODE REVIEW

**Comment by smichr** [comment]
You should probably factor out `_term_factors(f.expand()))` so you don't have to keep recomputing it.

**Comment by MaqOwais** [comment]
> You should probably factor out _term_factors(f.expand())) so you don't have to keep recomputing it.

Are you suggesting not to use this ` _term_factors(f.expand())) ` multiple times (in if elif statements ) and try to compress it in another way?

**Comment by oscarbenjamin** [comment]
Can you respond to this: https://github.com/sympy/sympy/pull/20613#issuecomment-752978781

The changes there as merged are not acceptable because they return incomplete solution sets over the complexes and the modifications to the ode code are hacky and should be removed.

If you can fix those things in this PR then great. Otherwise I am going to open a PR that reverts the changes in #20613.

**Comment by MaqOwais** [comment]
> If you can fix those things in this PR then great. Otherwise I am going to open a PR that reverts the changes in #20613.

I'll definitely try to see more into it.

File: sympy/solvers/bivariate.py [hunk 1]
```diff
@@ -126,6 +138,8 @@ def _lambert(eq, x, domain=S.Complexes):
     where X = g(x) and x = g^-1(X), return the Lambert solution,
         ``x = g^-1(-c/b + (a/d)*W(d/(a*b)*exp(c*d/a/b)*exp(-f/a)))``.
     """
+    global lambertReal
+    lambertReal = None
```
Comment (oscarbenjamin) [inline sympy/solvers/bivariate.py:142]: No global variables

File: sympy/solvers/bivariate.py [hunk 2]
```diff
@@ -152,8 +166,24 @@ def _lambert(eq, x, domain=S.Complexes):
...
+        consistent_dom = consistent_domain(dom)
+
     xusolns = solve(X1 - u, x)
 
+    try:
+        if consistent_dom:
+            domain = S.Reals
+        else:
+            domain = S.Complexes
+    except UnboundLocalError:
```
Comment (oscarbenjamin) [inline sympy/solvers/bivariate.py:184]: We should never catch UnboundLocalError

**Comment by oscarbenjamin** [comment]
The flow control here is messy and needs to be simplified so that the code is better structured.

File: sympy/solvers/bivariate.py (hunk 1 above)
Comment (MaqOwais) [reply sympy/solvers/bivariate.py:142]: is there any other means to access variables across the functions so that to avoid global variables?

Comment (oscarbenjamin) [reply sympy/solvers/bivariate.py:142]: The normal was is by using function parameters and arguments. I don't know what this variable is for but is it not just possible to check the `domain` argument?

Comment (MaqOwais) [reply sympy/solvers/bivariate.py:142]: >  I don't know what this variable is for but is it not just possible to check the domain argument?
This case is for if there are symbols in the equation but the lambert function is in a real range...

i.e.

```
b = Symbol('b')
eq = 3*log(a**(3*x + 5)) + b*log(a**(3*x + 5)) + a**(3*x + 5)
assert solveset_real(eq, x) == Intersection(FiniteSet(Rational(-5, 3) - LambertW(Rational(1, 3))/(3*log(p))), Interval(0, oo))

```
Here, ` LambertW(Rational(1, 3))` is real.

File: sympy/solvers/solvers.py [hunk 3]
```diff
@@ -376,7 +376,7 @@ def checksol(f, symbol, sol=None, **flags):
     # TODO: improve solution testing
 
 
-def solve(f, *symbols, **flags):
+def solve(f, *symbols, domain=S.Complexes, **flags):
```
Comment (oscarbenjamin) [inline sympy/solvers/solvers.py:379]: We don't want to add a `domain` argument to `solve`.

**Comment by oscarbenjamin** [comment]
This PR seems to have gone off on a big tangent. We do not want to change `solve` by adding a `domain` argument.

You need to understand that the `solveset` code should not call `solve`. The `solve` function is a legacy function and the `solveset` function is intended to provide a more mathematically correct replacement with a more well-defined interface.

The basic problem with #20613 is that it is using code from `solve` as part of `solveset`.

I will revert #20613 until we can come up with something clean for `solveset`.

**Comment by oscarbenjamin** [comment]
I've opened a PR #20814 to revert #20613

**Pull request closed without merging**